File: src/p0005/cpp/solution.cpp
class Solution {
public:
    string longestPalindrome(const string &s) {
        auto n = 2 * s.length();
        auto asymmetric = [&s, n](size_t center, size_t radius) {
            return center < radius || center + radius > n || (center % 2 != radius % 2
                            && s[(center - radius) / 2] != s[(center + radius) / 2]);
        };
        size_t preCenter = 0, preRadius = 0, maxCenter = 0, maxRadius = 0;
        vector<size_t> p(n, 0);
        for (size_t i = 1; i < n; i++) {
            auto distance = i - preCenter; if (distance <= preRadius) {
                p[i] = min(p[preCenter - distance], preRadius - distance);
            }
            auto curCenter = i, &curRadius = p[i];
            while (!asymmetric(curCenter, curRadius + 1)) {
                curRadius += 1;
            }
            if (preCenter + preRadius < curCenter + curRadius) {
                preCenter = curCenter; preRadius = curRadius;
            }
            if (maxRadius < preRadius) {
                maxCenter = preCenter; maxRadius = preRadius;
            }
        }
        return s.substr((maxCenter - maxRadius) / 2, maxRadius);
    }
};

File: src/p0005/python/solution.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = 2 * len(s)

        def asymmetric(center, radius):
            return center < radius or center + radius > n or (center % 2 != radius % 2
                            and s[(center - radius) // 2] != s[(center + radius) // 2])

        pre_center, pre_radius, max_center, max_radius = 0, 0, 0, 0
        p = [0] * n
        for i in range(1, n):
            distance = i - pre_center
            if distance <= pre_radius:
                p[i] = min(p[pre_center - distance], pre_radius - distance)
            while not asymmetric(i, p[i] + 1):
                p[i] += 1
            if pre_center + pre_radius < i + p[i]:
                pre_center, pre_radius = i, p[i]
            if max_radius < pre_radius:
                max_center, max_radius = pre_center, pre_radius
        return s[(max_center - max_radius) // 2: (max_center + max_radius) // 2]
